packages/rewriter/lib/threads/index.ts

```diff
@@ -0,0 +1,242 @@
+import _ from 'lodash'
+import Bluebird from 'bluebird'
+import Debug from 'debug'
+import * as path from 'path'
+import os from 'os'
+import { MessageChannel, Worker } from 'worker_threads'
+import { RewriteRequest, RewriteResponse } from './types'
+import { DeferSourceMapRewriteFn } from '../js'
+
+const debug = Debug('cypress:rewriter:threads')
+
+const _debugWorker = !debug.enabled ? _.noop : (worker: WorkerInfo) => {
+  return { ..._.pick(worker, 'isBusy', 'id'), freeWorkers: _.filter(workers, { isBusy: false }).length }
+}
+
+const _debugOpts = !debug.enabled ? _.noop : (opts: RewriteOpts) => {
+  return { ..._.pick(opts, 'isHtml'), sourceLength: opts.source.length }
+}
+
+// in production, it is preferable to use the transpiled version of `worker.ts`
+// because it does not require importing @packages/ts like development does.
+// this has a huge performance impact, bringing the `responsiveMs` for threads
+// from ~1s to about ~300ms on my system
+const WORKER_FILENAME = process.env.CYPRESS_INTERNAL_ENV === 'production' ? 'worker.js' : 'worker-shim.js'
+
+const WORKER_PATH = path.join(__dirname, WORKER_FILENAME)
+
+const MAX_WORKER_THREADS = Math.max(8, os.cpus().length)
+
+type DeferredPromise<T> = { p: Promise<T>, resolve: () => {}, reject: () => {} }
+
+type WorkerInfo = {
+  id: number
+  thread: Worker
+  isBusy: boolean
+}
+
+type QueuedRewrite = {
+  deferred: DeferredPromise<string>
+  opts: RewriteOpts
+}
+
+type RewriteOpts = Pick<RewriteRequest, 'url' | 'source' | 'isHtml' | 'sourceMap' | 'inputSourceMap'> & {
+  deferSourceMapRewrite?: DeferSourceMapRewriteFn
+}
+
+const workers: WorkerInfo[] = []
+const queued: QueuedRewrite[] = []
+
+let originalProcessExit
+
+// HACK: electron can SIGABRT if exiting while worker_threads are active, so overwrite process.exit
+// to ensure that all worker threads are killed *before* exiting.
+// @see https://github.com/electron/electron/issues/23366
+function wrapProcessExit () {
+  if (originalProcessExit) {
+    return
+  }
+
+  originalProcessExit = process.exit
+
+  // note - process.exit is normally synchronous, so this could potentially cause strange behavior
+  // @ts-ignore
+  process.exit = _.once(async (...args) => {
```

Comment:
make this less hacky

1. create process.exitAsync with clearly defined semantics
2. make process.exit throw an error so we don't accidentally use it

Reply:
going to leave this in for now and prioritize cleaning it up before taking this out of experimental. it would be far preferable to fix this in Electron than have to hack around it ourselves, so let's hold off for now.